spring-kafka/src/test/java/org/springframework/kafka/annotation/BatchListenerConversionTests.java

```diff
@@ -134,24 +140,55 @@ public StringJsonMessageConverter converter() {
 		}
 
 		@Bean
-		public Listener listener() {
-			return new Listener();
+		public Listener listener1() {
+			return new Listener("blc1");
+		}
+
+		@Bean
+		public Listener listener2() {
+			return new Listener("blc2");
 		}
 
 	}
 
 	public static class Listener {
 
+		private final String topic;
+
 		private final CountDownLatch latch1 = new CountDownLatch(1);
 
-		private Object received;
+		private final CountDownLatch latch2 = new CountDownLatch(1);
+
+		private List<Foo> received;
+
+		private List<String> receivedTopics;
 
-		@KafkaListener(topics = "blc1")
-		public void listen(List<Foo> foos, @Header(KafkaHeaders.OFFSET) List<Long> offsets) {
-			this.received = foos;
+		private List<Integer> receivedPartitions;
+
+		public Listener(String topic) {
+			this.topic = topic;
+		}
+
+		@KafkaListener(topics = "#{__listenerBean__.topic}", groupId = "#{__listenerBean__.topic}.group")
```

Comment:
Can't advertise the feature like SpEL evaluation context variable and therefore register the outer object under a `#listener` name? More convenient to use and familiar syntax for SpEL users: https://docs.spring.io/spring/docs/5.0.2.RELEASE/spring-framework-reference/core.html#expressions-ref-variables ?

What I feel fragile for is that funky `beanRef` on the annotation...

Just my feeling from the end-user perspective.
If you are not agree with me, I'll merge it as you did.
Thanks

Reply:
That was my initial plan, but we don't have access to the SpEL evaluation context - it is a single instance that's cached in the `StandardBeanExpressionResolver` - it would require replacing all that resolution infrastructure with our own.

Reply:
Yes, I checked the code there, bu I mean *pretend* like we have it as a SpEL variable , by the name of the object in scope you use already.

Reply:
That works already, out of the box...

```java
@Bean
public Listener listener1() {
    return new Listener("topic1");
}
```

with

```java
    @KafkaListener(topics = "#{listener1.topic}",
        groupId = "#{listener1.topic}.group")
    public void listen(...) {
        ...
    }
```

But we're trying to solve a different problem, where we don't know the bean name while writing the code...

```java
@Bean
public Listener listener1() {
    return new Listener("topic1");
}

@Bean
public Listener listener2() {
    return new Listener("topic2");
}
```

We somehow need a reference to the current bean, but not by the bean name. Yes, ideally, a variable `#currentBean` would be best but, for the reason I specified above, we can't do that without copying a lot of code.

Reply:
OK. Let's go simpler discussion way!
I don't like the name for the object in `scope` as `__listenerBean__`. I would prefer `#listener` instead. This way it mimics the well-known SpEL variable syntax. 

Also I don't like that new `beanRef` annotation attribute. Don't know yet why, but my feeling like I would stay away of it. Although I agree about the argument "unlikely event that you have an actual bean called" especially in this point release. 

So, having all of this thinking in my head I only suggest to rename the default  `__listenerBean__` key to the `#listener`. And that's all.

Let me know if my point is still unclear!

Reply:
We can't use `#listener` because SpEL will think it's a real variable and fail to find it.

Reply:
I'd be ok with `_listener` or `_listener_` as the default.

Reply:
Ah! Good point!

OK. One more attempt. The `StandardBeanExpressionResolver` has this:
```
/**
 * Template method for customizing the expression evaluation context.
 * <p>The default implementation is empty.
 */
protected void customizeEvaluationContext(StandardEvaluationContext evalContext) {
}
```
Looks like this is the place where we could inject a SpEL variable for the bean in the method:
```
protected void processListener(MethodKafkaListenerEndpoint<?, ?> endpoint, KafkaListener kafkaListener, Object bean,
			Object adminTarget, String beanName) {
```
But I agree that is more coding, `ThreadLocal`, param propagation etc. And in the end technically we end up with the same logic for end user.

So, let me know if you would like to play with this or let's leave as is and merge!

Reply:
> I'd be ok with `_listener` or `_listener_` as the default.

Yeah... That really might be two `_` in the beginning, but I'm some lost to read dot after two `_`...

Reply:
ok; I will change it; and do no more; I think we've spent enough time on this 😄